feat: avatars pooling

pravusjif opened the pull request:
What this PR brings to the table:

- [x] Potential pooling support for all components
- [x] Pooling implemented for `AvatarShape`
- [x] Wearables are re-used (instead of being always instantiated). The material colors are applied after recycling and reverted on release.

Minor improvements:
- Mark green blocker pools as persistent, so the pool is never cleared up.
- Improvements on the `ParcelScenesCleaner` algorithm. Found a bug that made entities not being cleared up properly after teleports and could potentially destroy a scene out of nowhere.

More last minute fixes:
- fix: all avatar loading logic was being executed when updating any avatar with same wearables.
- fix: all avatar loading logic was being executed when performing emotes.
- fix: existing wearables rootBone was not being updated when changing bodyShape, leading to missing or floating wearables in random cases.
- fix: rootBone was being assigned needlessly on unequipped wearables when changing wearables.
- fix: now avatars can be loaded even when gameObjects are inactive--the loading will resume when the object is re-activated

pravusjif committed:
Added modified files for first implementation. The avatars pool is being initialized and used, but the avatars cleaning when recycling them is still WIP

pravusjif committed:
Removed avatars pool prewarm own-profile initialization as it wasn't needed

pravusjif committed:
Fixed problem with accumulated wearables and repositioning when recycling avatarshapes

Commit:
feat: add basic factory pool impl.

Commit:
- added concept of "Reset" vs "Cleanup" in pools. (cleanup is when resources are freed, reset is when resources are initialized)
- rename ResetAvatar() to CleanupAvatar() to reinforce this concept.

Commit:
- small fix to finally enable pools in avatars

Commit:
- add pooling support to ComponentUpdateHandler support
- ComponentUpdateHandler now uses CoroutineStarter

Commit:
- fix pink materials issue
- re-enabled forceNewInstance temporarily to avoid pink avatar issue

pravusjif committed:
Fixed pooled avatars name display

pravusjif committed:
Added ComponentUpdateHandler Cleanup() method to fix same-avatars-recycling from pool

Commit:
fix: entities pool prewarm had an unwieldy lifecycle. This could lead to force prewarm being not called at all when it should.
fix: add safe code to prevent prewarm multiple times for factories.
fix: nullRef in RemoveFromPool

Commit:
add test to reproduce avatar re-positioning issue

Commit:
Merge remote-tracking branch 'origin/fix/component-factory-pools' into fix/component-factory-pools

Commit:
fix: avatar not repositioning correctly on repool issue

Commit:
chore: remove unneeded logs

Commit:
refactor: add generic pool lifecycle interface and integrated to BaseComponent, so we have an easy way to integrate pools into components.

Commit:
fix: disable forceNewInstance on avatars

pravusjif committed:
Fixed pooled avatars positioning

pravusjif committed:
Merge branch 'fix/component-factory-pools' of https://github.com/decentraland/explorer into fix/component-factory-pools

pravusjif committed:
Added missing compiler conitional that broke the build

pravusjif committed:
Updated tests

pravusjif committed:
Updated tests

Commit:
refactor: move pool lifecycle handling to PoolableObject.cs

Commit:
Merge remote-tracking branch 'origin/feat/AvatarsPooling' into feat/AvatarsPooling

Commit:
Merge remote-tracking branch 'origin/master' into feat/AvatarsPooling

Commit:
fix: avatars using random recycled transforms for movement because the AvatarMovementController didn't recycle correctly

Commit:
test fixes

Commit:
Merge branch 'master' into feat/AvatarsPooling

Commit:
fixed failing builder test

Commit:
Merge remote-tracking branch 'origin/feat/AvatarsPooling' into feat/AvatarsPooling

Commit:
more test fixes

pravusjif committed:
removed debug.log

Commit:
code cleanup

Commit:
mark blockers pool as persistent

Commit:
remove null checks from getter

Commit:
code cleanup + add ParcelScenesCleaner fix to coroutine and ensure the scenes to be removed are not loaded

Commit:
Merge remote-tracking branch 'origin/feat/AvatarsPooling' into feat/AvatarsPooling

Commit:
fix: ensure minimap indicators are resetted correctly

Commit:
fix: remove old MinimapRepresentation from AvatarShape, name reset issue

pravusjif committed:
Fixed character preview renderers layer reset when exiting the Avatar HUD after changing gender

Commit:
Merge branch 'master' into feat/AvatarsPooling

Commit:
fix: scene unload test fail

sandrade-dcl reviewed (commented):
I'm trying to generate bots with `bots.create()` but I can't because it is throwing a null reference exception in the `UpdateFromJSON()` (BaseComponent, line 63). It seems the `updateHandler` variable comes null.

Edit: Ok, it was a problem in my local project. I regenerated the Library folder and it works.

moliva committed:
fix: allow users to wear exclusive wearables on test mode

pravusjif committed:
toggled forceNewInstance to true in WearableController temporarily

pravusjif committed:
fixed possible invisible avatar when confirming avatar editor mid-loading

BrianAmadori commented on unity-client/Assets/Scripts/MainScripts/DCL/Controllers/HUD/AvatarEditorHUD/Scripts/AvatarEditorHUDView.cs:
```diff
@@ -300,6 +303,8 @@ public void SetVisibility(bool visible)
         avatarEditorCanvasGroup.blocksRaycasts = visible;
         isOpen = visible;
+
+        // characterPreviewController.gameObject.SetActive(visible);
```
Remove commented line

pravusjif committed:
removed commented code

Commit:
fix: avatars were being re-loaded on any profile update, even if it contained no avatar changes
fix: old wearables were not unloaded correctly upon wearable or sex change, causing undefined behaviour

Commit:
Merge remote-tracking branch 'origin/feat/AvatarsPooling' into feat/AvatarsPooling

pravusjif committed:
added done button locking when loading the character preview avatar in avatar HUD

pravusjif committed:
Toggled forceNewInstance to false again

Commit:
refactor: AvatarRenderer logic overhaul

- fix: existing wearables rootBone not being updated when changing bodyShape, leading to missing or floating wearables in random cases
- fix: all avatar loading logic being executed when performing emotes
- fix: rootBone being assigned needlessly on unequipped wearables when changing wearables
- fix: all avatar loading logic being executed when updating avatar with same model
- fix: now avatars can be loaded even when gameObjects are inactive--the loading will resume when the object is re-activated

Commit:
Merge branch 'feat/AvatarsPooling' of github.com:decentraland/explorer into feat/AvatarsPooling

sandrade-dcl reviewed (approved):
**Only as annotation for the future improvements:** I have noticed the framerate drop that we have while we are with a large number of avatars in the scene (from 60fps to 35-40 aprox.), is only happening when the camera is rendering the avatars. In fact if we look at other place where there are not avatars (even with all these avatars near of us), the framerate go up to 60fps.

Commit:
test fixes
